### ISSUE TITLE
`moveit`: allow C++ objects to be owned by value in Rust

### ISSUE THREAD

**Issue by adetaylor** [google/autocxx#379]
We should see if we can integrate this:
https://crates.io/crates/moveit
instead of automatically wrapping and unwrapping non-trivial C++ objects into `UniquePtr`. This shouldn't really make any significant difference to ergonomics, but should remove the overhead of the `std::move`s into and out of heap cells.

**Comment by ssbr** [comment]
tl;dr: I think moveit can't help with the API surfaces, but only helps with implementing functions and taking values to their final storage location.

(Caveat: I have never written code using `Pin` and have never used moveit, so I'm gonna ask mcyoung to review this reply.)

---

This only matters, as pointed out, when Rust needs to obtain or give up ownership of a C++ object (e.g. CxxString). In C++, there's a handful of interesting patterns where we might expect Rust to be on either side of the call:

```c++
// Acquires permanent ownership of a string, and takes it by value so that it can be done via efficient move rather than copy.
void TakesStringByValue(std::string);

// "might" acquire permanent ownership of the argument, but also might not -- rvalue references don't necessarily have to result in an actual C++ move. Similar to above, this doesn't force a copy.
// (This could also take a std::string& or std::string*, but that would be less idiomatic and not work with temporaries on the part of the C++ caller.)
void TakesStringByRvalueReference(std::string&&);

// "might" keep a copy of the string, because anything that takes a reference might do that if it wants to.
void TakesStringByConstReference(const std::string&);
```

(see e.g. https://abseil.io/tips/117 for discussion about those first two APIs.)

Of these three: the first cannot be native in Rust, because a by-value `CxxString` *needs* to be trivially relocatable. moveit gets around this by only exposing references, so it would still need to be a reference in Rust land. In particular, I think there is no escaping that we will be receiving a `Pin<&mut CxxString>`. The second doesn't exist in Rust, but is morally similar to a `Pin<&mut CxxString>` -- rvalue references are very similar to regular references. The third already works as `&CxxString`.

Similarly, on the *caller* side, we still cannot provide a native `CxxString` value, and the latter two still must be some kind of reference.

So I think there's no getting around the fact that, within Rust, rvalue references will need to either be `Pin<&mut T>` or some kind of wrapper type like `Pin<RvalueReference<T>>` (or maybe just `RvalueReference<T>`?).

And more significantly, I think there's no getting around the fact that "by-value" objects will *also* need to be references in Rust, in particular `Pin<&mut T>` or some wrapper thereof, and actual move-construction can only ever happen on the C++ side of the ffi layer, or at a delayed point inside the Rust code via moveit. In particular, just like you could have `Pin<RvalueReference<T>>`, you could have, idk, `Pin<LazyValue<T>>`: on the C++ side, this is represented as a by-value T. On the Rust side, this is merely a reference to T (the *owned* copy only ever materializes on the C++ end), and you can move it again via `moveit`.

So what `moveit` gives us is the ability to, once we have a `Pin<&mut T>`, actually create a copy on the Rust side, and do it via move-construction. But I don't think it can help improve the API surface itself, exactly because we can't avoid using `Pin<&mut T>` at the API boundaries. And in fact, we can do the `Pin<&mut T>` kind of API even without moveit, moveit just makes the implementations more efficient by allowing us to move inside them, rather than only copy.

**Comment by ssbr** [comment]
I talked _briefly_ with Miguel (moveit author) and he pointed out that I'm missing the destructive aspect of moveit: in particular, we can't just use `Pin<&mut T>`, because this doesn't let us destroy the T, and moveit requires being able to destroy the original value.

I think what I wrote above, therefore, is only really correct for `T&&`, which moveit can't move (because it can't destroy).

For by-value `T`, we _could_ do it that way, at the cost of not being natively interoperable with moveit, and requiring an extra step to get from a `Pin<&mut T>` to a `StackBox` or similar that moveit can understand. But given that we needed that anyway for `T&&`, maybe not a huge loss?

(I confess I am not sure why moveit wants to destroy things! We definitely need a move-without-destroy trait for T&&, right?)

---

The moveit-aware way for passing `T` by value, AIUI, would be: when calling Rust from C++, store in a `StackBox`. This changes who destroys the object from C++ (at the end of the stack frame) to Rust (when it moves). When calling C++ from Rust, I thiiiink the moveit-native way would be to use a Ctor, but again this is only moving when the destructor is invoked from the caller-site to the callee-site. In either case the destructor is called all the same, right?

(I'm out of my element for sure, so I'm gonna ask Miguel to reply on Monday.)

**Comment by adetaylor** [comment]
Here are some notes about how we'd tackle integrating this into `autocxx`, assuming that `moveit` _can_ be used to transfer objects over the FFI boundary (there's debate happening in the `moveit` repo).

# Policy decisions

First of all, there's no point doing this until and unless `moveit` can allow objects to be passed by value across the FFI boundary. There is limited utility in allowing C++ objects to be held by value in Rust otherwise.

Second, at the moment autocxx gives users a choice for each given C++ type:
* `generate!` - makes an opaque type which is automatically boxed into `std::unique_ptr` - corresponds to `ExternType::Opaque` in `cxx` terminology
* `generate_pod!` - passes the struct directly - corresponds to `ExternType::Trivial` in `cxx` terminology. `cxx` confirms via static assertions that such a type has a trivial move constructor and no destructor.

(#17 notes that this terminology may not be ideal).

We need to decide if we want to replace the meaning of `generate!` such that _all_ types can be held by value in Rust, or if we want a _third_ option. Obviously if we _can_ use `moveit` for all non-POD types, that's ideal, but I have doubts about our ability to represent some fields (see below).

# Implementing the necessary traits

`conversion/analysis/pod/mod.rs` function `analyze_pod_api` either:
* Makes a type opaque if we can't promise it's plain old data (i.e. no move constructor, etc.) - the function `make_non_pod`. or
* Notes all the field types

`make_non_pod` should change. We would want to:
* Keep the fields (where possible)
* Implement the `moveit` traits

We would also very likely want to add some more items to the `extra_apis` out parameters, representing the functions we need to call for each move/copy constructor. These probably have new `ApiDetail` types called `MoveConstructor` and `CopyConstructor` - but see the next section, it might be easier to make them a special case of `ApiDetail::Function`.

# Generating move constructors and copy constructors

Those new `Api`s with `ApiDetail::MoveConstructor` and `ApiDetail::CopyConstructor` will be treated exactly like any `ApiDetail::Function` during subsequent analysis stages. In fact, we might well want to make them a special case of `ApiDetail::Function` (there are already all sorts of special cases for constructors, methods, static methods etc.)

After the various analysis phases, those new `Api`s will make it into `codegen_rs` and `codegen_cpp`, which can generate the Rust and C++ side code for each move/copy constructor, respectively.

Another alternative is to assume that _any_ non-POD `ApiDetail::Type` requires such move/copy constructors to be generated, and therefore there's no need to add any new `Api`s at all. However, that feels like it would require more work to persuade the `conversion/analysis/fun` code to do the right thing. All in all, making these things special cases of `FnKind` seems list the best bet.

# Applying suitable conversions to types as they're passed across the FFI boundary

* `conversion/fun/mod.rs` has a function `argument_conversion_details` which returns a `TypeConversionPolicy`. Similarly, there's `return_type_conversion_details`. Right now, these know about ~three policies:
  * For types safe to represent by value - no conversion is needed.
  * For others, instead require the type be packaged into a `std::unique_ptr` on the C++ side. This is the `TypeConversionPolicy` which needs to change.
  * For strings, there's special handling but it's basically the same.
* This boils down to the following policy:
```rust
        TypeConversionPolicy {
            unwrapped_type: ty,
            cpp_conversion: CppConversionType::FromUniquePtrToValue,
            rust_conversion: RustConversionType::None,
        }
```
- it should be pretty obvious that we want new `CppConversionType` and `RustConversionType` enum variants.
* Then it should really just be a matter of implementing those types of conversion in the Rust and C++ code generators, which will automatically fail to compile until you've done so.
* I assume that move/copy constructors won't do any of this fanciness, so any extra conversion policy will be required only if we need to do something special as a `moveit`-compatible type is passed over the FFI boundary.

# Troublesome fields

Currently, we remove all fields from non-POD structs. This saves us from having to deal with complicated, nasty field types, because we just treat `struct`s as opaque under nearly all circumstances.

If `moveit` causes us to keep all the fields intact, we're going to run into problems with a wider variety of types.

There are situations in which `bindgen` itself gives up and declares a type opaque, and we obviously follow its lead in such cases.

We may well need to fall back to truly opaque types in many circumstances.

# Accessors

#53 says we don't currently provide accessor methods to opaque types. `moveit` might resolve that, if we really are able to keep the fields.

# Strings

Right now we use `cxx::CxxString` which:
* obviously doesn't impl all the required `moveit` traits, but...
* ... has a lovely ergonomic API which is known to `cxx` users.

It's a shame, but we'd probably need to have our own `CxxString` equivalent type, unless `moveit` turns out to be so amazing that we can get it adopted upstream in cxx.

Anyway, there's some notes about what we'd need to do. I am highly unlikely to get to this any time soon, but I'm leaving my notes in case anyone else wants to have a crack at it :)

**Comment by adetaylor** [comment]
Phasing of work here.

[] Double check that opaque types we make are `!Unpin`
[] Add a `feature` to `autocxx` for `moveit` support
[] Improve CI so that we do a build/test run with `moveit` enabled
[] Write a test which aims to use `moveit::new::New` to construct a C++ object in place
[] Implement the trait `moveit::new::New` for any regular constructors, using placement new on the C++ side into a pointer passed from the Rust side
[] Do the same for `MoveNew` where we detect move constructors (or specifically, where we _don't_ detect deleted move constructors)
[] Do the same for `CopyNew`

I'm a bit hazy from here on.

[] Investigate what exactly `ExternType::Trivial` does in `cxx`.
[] Switch all types to be `Trivial`, but, ensure they can't be created except in a pinned way...? Check if that's OK.
[] Argument conversion for calls into C++ functions: where we previously required the Rust-side type `UniquePtr<T>`, instead we take `&T` (or maybe `Pin<&mut T>`). We know `T: CopyNew` because we implemented that trait for `T`, but we won't actually need that. The C++ wrapper function, instead of unboxing a `UniquePtr`, simply uses `*the_reference_passed_from_Rust`.
[] Argument conversion for return values from C++ functions: previously we boxed them into a `UniquePtr`. Instead now we... just return them directly?
[] Recognize functions taking/returning `&&` and do appropriate shenanigans with `MoveNew` instead.
[] Consider adding direct field access.

**Comment by adetaylor** [comment]
Progress report on this. I haven't been doing it exactly according to the phasing scheduled above.

* We now generate implicit constructors (#760). This is useful even without `moveit`. One currently known bug, #815.
* We implement `moveit::New` for any constructors, `Drop` for destructors, and `moveit::CopyNew`, `moveit::MoveNew` for any copy and move constructors. This educates Rust about what types are movable and copyable, allowing these to be used in trait bounds. Most importantly, it allows such types to be created on the stack using the `moveit!` macro, which means methods can be called on them and references to them can be passed into other C++ functions.
* Implement `UniquePtr::emplace<T: New>` (#767)

Done but not yet landed:
* Allow moving _out of_ a `UniquePtr<T>` and onto the stack (#766 and related moveit change)
* Alter C++ functions which take a parameter by value such that they can either consume a `UniquePtr<T>` or take a `T&` if `T: CopyConstructor`. This happily seems to be a terrific compromise between normal Rust semantics (consume) and normal C++ semantics (copy) - we can support either, ergonomically (#766)
* Discard all the complex code which implements `make_unique` and replace it with something simpler that just calls `UniquePtr::emplace` (#742)

Still to do but well understood:
* Start to support rvalue parameters (should be simple after #766 lands)
* Implement `Clone` for a `UniquePtr<T: CopyNew>` (actually it would have to be a new trait due to orphan rule)

Decisions still pending:
* Return values. Ideally with `moveit` we return an `impl New`, but this is unergonomic, so for now I'm still returning a `UniquePtr`. That means people can't keep return values on the stack, though. Decision needed. It may be that it's reasonably ergonomic with modern moveit.
* All the fiddly bits listed at the end of the previous long comment!
* Cease to call `std::move` deep in the generated code that handles a `ValueParam`. We should be able to pass pointers straight through, removing one copy and avoiding slightly weird semantics here.

**Comment by adetaylor** [comment]
I consider this essentially now done.
